expression/builtin_time_vec.go

```diff
@@ -1744,11 +1744,71 @@ func (b *builtinToDaysSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column)
 }
 
 func (b *builtinDateFormatSig) vectorized() bool {
-	return false
+	return true
 }
 
+// vecEvalString evals a builtinDateFormatSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_date-format
 func (b *builtinDateFormatSig) vecEvalString(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	n := input.NumRows()
+	buf, err := b.bufAllocator.get(types.ETDatetime, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+	if err := b.args[0].VecEvalTime(b.ctx, input, buf); err != nil {
+		return err
+	}
+
+	buf0, err := b.bufAllocator.get(types.ETString, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf0)
+	if err := b.args[1].VecEvalString(b.ctx, input, buf0); err != nil {
+		return err
+	}
+
+	buf1, err := b.bufAllocator.get(types.ETInt, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf1)
+	if err := b.args[0].VecEvalInt(b.ctx, input, buf1); err != nil {
+		return err
+	}
+	result.ReserveString(n)
+	ds := buf.Times()
+	i64s := buf1.Int64s()
+	for i := 0; i < n; i++ {
+		if buf.IsNull(i) || buf0.IsNull(i) {
+			result.AppendNull()
+			continue
+		}
+		if buf0.GetString(i) == "0" {
+			result.AppendString("0")
+			continue
+		}
+		str, err := ds[i].DateFormat(buf0.GetString(i))
+		if ds[i].InvalidZero() {
+			if err != nil {
+				return err
+			}
+			result.AppendString("")
+			flag := 1
+			if err := handleInvalidTimeError(b.ctx, types.ErrIncorrectDatetimeValue.GenWithStackByArgs(ds[i].String())); err != nil {
+				flag = 0
+			}
+			isOriginalIntOrDecimalZero := i64s[i] == 0 && flag == 1
```

Comment:
In non-vec function, `isOriginalIntOrDecimalZero := n == 0 && !isNullInt && errInt == nil`. But in this pr, you use the `flag` represent `!isNullInt && errInt == nil`?  I think the behavior is different from non-vec.